ConditionalCompilation.md

```diff
@@ -1,28 +1,140 @@
-  * More structured: ties to source elements not lines
-    * Perhaps some optimization potential..
-      * For example could parse/lower certain functions ahead of 
-        time and easily remove them from the IR vs #if #else which operates on a source level.
-    * Is it an wgsl attribute? syntax makes sense but wgsl 
-* what are some examples of things to conditionally compile?
-  * function variants e.g. select which function variant to use based on condition
-  * struct fields e.g. include fields only if condition is set
-  * import statements? e.g. import util/fast vs util/regular based on condition
-* how do we support build time vs runtime linking?
-  * e.g. some projects will want prebuilt versions for various mobile configurations.
-    * ref [unity shader conditionals](https://docs.unity3d.com/Manual/shader-conditionals.html)
-  * should the wesl code distinguish conditions that are settable at compile time from those that are settable at runtime?
+## Overview
+
+Conditional compilation is a mechanism to modify the source code based on parameters passed to the compiler. We distinguish two kinds:
+
+ * **unstructured**: arbitrary code sections can be injected or conditionaly included. This is what the C preprocessor and macros do, and a lesser version of that is proposed in [`simple templating`](SimpleTemplating.md). 
+ * **structured**: only structural elements of the syntax can be manipulated, e.g. a whole declaration, a member, etc. Rust uses this approach with the `#[cfg]` attribute.
+
+We think that *structured* is the way to go. It leads to clearer and safer code, which is more important than implementation complexity in our eyes.
+Also, a language that has a good expressive power already should not need a way to hack around with arbitrary code injection, like C macros do.
+
+| unstructured | structured |
+|--------------|------------|
+| (+) much easier to implement: just look for the `#word` symbols | (-) harder to implement: requires a full parsing |
+| (+) often language-agnostic (see the [C preprocessor](https://en.wikipedia.org/wiki/C_preprocessor). Familiar to C developers. | (-) a new syntax needs to be taught, not always self-explanatory |
+| (-) poorly integrated in the language, harder to read by humans | (+) is a well-designed syntax feature of the language |
+| (+) technically more expressive (e.g. [manipulating identifiers](https://en.wikipedia.org/wiki/C_preprocessor#Token_concatenation)) | (-) can only conditionaly include parts of the syntax tree, poor text generation capabilities. |
+| (-) behaves unpredictably, intent and behavior is hidden | (+) intent and behavior is visible at the usage site |
+| (-) no type checking, no syntax checking, because code is generated dynamically | (+) can statically check that conditional variants lead to syntactically correct and type-valid code |
+| (-) poor IDE support | (+) IDE can check all possible code paths |
+
```

Comment:
Excellent discussion. Though (as you reminded me on a recent pr!) I think it should move to a separate location so folks can visit the spec and not have to read the back story unless they’re interested.

Reply:
What enables the IDE to know the valid combinations of conditions? 

And if there is a way to know the valid combinations of conditions, why is the siutation different for structured vs. unstructured?

Reply:
valid combinations cannot be checked. This is the case e.g. in rust too.
I'll clarify this.

Reply:
To answer that: With a structured `@if`, one can always generate a unified syntax tree

```
struct Foo {
  a: f32,
  @if(some_condition)
  b: f32
}
```
would turn into a syntax tree along the lines of
```
(struct foo (member a) (if some_condition (member b)))
```

This means that at the very least, a linker can always check if the WGSL code is syntactically (not semantically) valid.

Meanwhile with `#ifdef`, one can frequently end up with multiple separate syntax trees. The above example could be turned into a single syntax tree. Meanwhile
```
struct FooBar { // Which bracket is the closing bracket?
  a: f32,
#ifdef some_condition
}
#else
  b: f32
}
#endif
```
is *very* hard to turn into a single syntax tree. Juggling two separate syntax trees in a linker or language server is a whole lot of extra effort, so nobody does it.
This example in particular also breaks one IDE feature: Jump-to-matching-bracket


And finally, `@if` is equally powerful as `#ifdef`. Proof:
- Assume all combinations of conditions are valid
- Expand the `#ifdef` into all of the combinations
- Encode each combination in a separate `@if`

The above `FooBar` example could thus be naively translated as
```
@if(!some_condition)
struct FooBar { 
  a: f32,
}
@if(some_condition)
struct FooBar { 
  a: f32,
  b: f32
}
// Oh look, two syntactically valid WGSL structs. Happiness :)
```
Yes, that's code duplication, but it's this example is just there to illustrate the fact that `@if` is equally powerful.

Reply:
+1 for providing a proof